# Movies Tonight Part 3: SQL DML

In [1]:
%load_ext sql
%sql sqlite:///Movies.db

'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

In [46]:
%%sql
SELECT * FROM DATASET LIMIT 10;
PRAGMA table_info(CREDITS);

 * sqlite:///Movies.db
Done.
Done.


[(0, 'CID', 'INTEGER', 0, None, 1),
 (1, 'MID', 'INTEGER', 1, None, 0),
 (2, 'AID', 'INTEGER', 1, None, 0),
 (3, 'CCODE', 'TEXT', 1, None, 0)]

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [55]:
%%sql
DELETE FROM MOVIES;
INSERT INTO MOVIES (Title,Rating)
    SELECT DISTINCT MTitle, Rating FROM DATASET;
SELECT * FROM MOVIES;

DELETE FROM THEATERS;
INSERT INTO THEATERS (Name,Location,Phone)
    SELECT DISTINCT TName, Location, Phone FROM DATASET;
SELECT * FROM THEATERS;

DELETE FROM ARTISTS;
INSERT INTO ARTISTS (Name)
    SELECT DISTINCT CName FROM DATASET;
SELECT * FROM ARTISTS;

/* We can work with this data in this way because there are no duplicate values on 'name' */

DELETE FROM SHOWS;
INSERT INTO SHOWS (SHOWTIME, MID, TID)
    SELECT DISTINCT DATASET.ShowTime, MOVIES.MID, THEATERS.TID
    FROM DATASET
        JOIN MOVIES ON (MOVIES.Title = DATASET.MTitle)
        JOIN THEATERS ON (THEATERS.Name = DATASET.TName);
SELECT SHOWTIME, TITLE, NAME FROM SHOWS JOIN MOVIES USING (MID) JOIN THEATERS USING (TID);

DELETE FROM CREDITS;
INSERT INTO CREDITS (CCODE, MID, AID)
    SELECT DISTINCT CCode, MID, AID
    FROM DATASET
        JOIN ARTISTS ON (ARTISTS.NAME = DATASET.CName)
        JOIN MOVIES ON (MOVIES.TITLE = DATASET.MTitle);
SELECT * FROM CREDITS
LIMIT 10;

 * sqlite:///Movies.db
23 rows affected.
23 rows affected.
Done.
9 rows affected.
9 rows affected.
Done.
153 rows affected.
153 rows affected.
Done.
Done.
131 rows affected.
Done.
162 rows affected.
162 rows affected.
Done.


[(1, 1, 1, 'A'),
 (2, 1, 2, 'A'),
 (3, 1, 3, 'A'),
 (4, 1, 4, 'A'),
 (5, 1, 5, 'A'),
 (6, 1, 6, 'A'),
 (7, 1, 7, 'A'),
 (8, 1, 8, 'A'),
 (9, 1, 9, 'D'),
 (10, 2, 10, 'A')]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

In [ ]:
YOUR CODE HERE

4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [ ]:
YOUR CODE HERE

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".

In [ ]:
Completed Part 3